core/transformer/src/IModelTransformer.ts

```diff
@@ -1132,10 +1133,44 @@ export class IModelTransformer extends IModelExportHandler {
     return Semver.gt(`${schemaKey.version.read}.${schemaKey.version.write}.${schemaKey.version.minor}`, Schema.toSemverString(versionInTarget));
   }
 
+  private _longNamedSchemasMap = new Map<string, string>();
+
   /** Override of [IModelExportHandler.onExportSchema]($transformer) that serializes a schema to disk for [[processSchemas]] to import into
-   * the target iModel when it is exported from the source iModel. */
-  public override async onExportSchema(schema: ECSchemaMetaData.Schema): Promise<void> {
-    this.sourceDb.nativeDb.exportSchema(schema.name, this._schemaExportDir);
+   * the target iModel when it is exported from the source iModel.
+   * @returns {Promise<ExportSchemaResult>} Although the type is possibly void for backwards compatibility of subclasses,
+   *                                        `IModelTransformer.onExportSchema` always returns an[[IModelExportHandler.ExportSchemaResult]]
+   *                                        with a defined `schemaPath` property, for subclasses to know where the schema was written.
+   *                                        Schemas are *not* guaranteed to be written to [[IModelTransformer._schemaExportDir]] by a
+   *                                        known pattern derivable from the schema's name, so you must use this to find it.
+   */
+  public override async onExportSchema(schema: ECSchemaMetaData.Schema): Promise<void | ExportSchemaResult> {
+    const ext = ".ecschema.xml";
+    let schemaFileName = schema.name + ext;
+    // many file systems have a max file-name/path-segment size of 255, so we workaround that on all systems
+    const systemMaxPathSegmentSize = 255;
+    if (schemaFileName.length > systemMaxPathSegmentSize) {
+      // this name should be well under 255 bytes
+      // ( 100 + (Number.MAX_SAFE_INTEGER.toString().length = 16) + (ext.length = 13) ) = 129 which is less than 255
+      // You'd have to be past 2**53-1 (Number.MAX_SAFE_INTEGER) long named schemas in order to hit decimal formatting,
+      // and that's on the scale of at least petabytes. `Map.prototype.size` shouldn't return floating points, and even
+      // if they do they're in scientific notation, size bound and contain no invalid windows path chars
+      schemaFileName = `${schema.name.slice(0, 100)}${this._longNamedSchemasMap.size}${ext}`;
```

Comment:
Is it guaranteed here that `schema.name` is not already present in `_longNamedSchemasMap`?

Reply:
yes, although not directly checked for. It's not possible to have schemas of the same name in the source, and that map is cleared before and after calls to `processSchemas`, (although the transformer is not actually designed for multiple uses anyway)